### ISSUE TITLE
Reporting an issue with the legacy Python 2.7 runtime files?

### ISSUE THREAD

**Issue by aetherknight** [GoogleCloudPlatform/appengine-python-standard#26]
I don't think this is necessarily the right place to report this issue, but I don't know where else to report a bug in the AppEngine standard files installed through the GCloud SDK. I have been talking with GCP support for a week, but they have not yet figured out who to forward the problem to. 

It also seems like this issue is related to the work being done in this project.

This issue is about the SDK for developing Python 2.7 AppEngine Standard projects. My team is well aware that Python 2.7 has reached EOL and we are working to migrate to Python 3, but GCP continues to support the legacy Python 2.7 runtime.

## Expected Behavior

The Google Cloud SDK, v358.0.0, contains an update to the `app-engine-python` and `app-engine-python-extras` modules that breaks the `testbed`/`api_server` functionality provided with the legacy AppEngine runtime for Python 2.7.

The following works with Google Cloud SDK v357.0.0, which contains `app-engine-python` and `app-engine-python-extras` v1.9.93, but it raises an error with GCloud SDK v358.0.0 and v359, which contains v1.9.94 of the AppEngine components.

Under Python 2.7, the following sample code should run without raising any exceptions:

```python
import sys
PATH_TO_GOOGLE_CLOUD_SDK = "/Users/wsorvis/google-cloud-sdk" # You will have to update this path
APP_ENGINE_RUNTIME = PATH_TO_GOOGLE_CLOUD_SDK + "/platform/google_appengine"
sys.path.insert(0, APP_ENGINE_RUNTIME)
import dev_appserver
dev_appserver.fix_sys_path()

from google.appengine.ext import testbed
mytestbed = testbed.Testbed()
mytestbed.activate(use_datastore_emulator=True)
mytestbed.init_datastore_v3_stub()
```

Again, this works as expected with Google Clou dSDK v358 (and v1.9.93 of the old AppEngine SDK files)

## Actual Behavior

However, after updating to Google Cloud SDK v358, it raises the following exception:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/wsorvis/google-cloud-sdk/platform/google_appengine/google/appengine/ext/testbed/__init__.py", line 640, in init_datastore_v3_stub
    delegate_stub.Clear()
  File "/Users/wsorvis/google-cloud-sdk/platform/google_appengine/google/appengine/ext/remote_api/remote_api_stub.py", line 670, in Clear
    self._server.Send('/clear?service=datastore_v3')
  File "/Users/wsorvis/google-cloud-sdk/platform/google_appengine/google/appengine/tools/appengine_rpc.py", line 486, in Send
    f = self.opener.open(req)
  File "/Users/wsorvis/.pyenv/versions/2.7.18/lib/python2.7/urllib2.py", line 435, in open
    response = meth(req, response)
  File "/Users/wsorvis/.pyenv/versions/2.7.18/lib/python2.7/urllib2.py", line 548, in http_response
    'http', request, response, code, msg, hdrs)
  File "/Users/wsorvis/.pyenv/versions/2.7.18/lib/python2.7/urllib2.py", line 473, in error
    return self._call_chain(*args)
  File "/Users/wsorvis/.pyenv/versions/2.7.18/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/Users/wsorvis/.pyenv/versions/2.7.18/lib/python2.7/urllib2.py", line 556, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
urllib2.HTTPError: HTTP Error 500: Internal Server Error
```

## Steps to Reproduce the Problem

1. Install Google Cloud SDK v357, and both of `app-engine-python` and `app-engine-python-extras` components 
2. Run the sample test code above with Python 2.7, observe that it does not raise an exception and it returns a 0 status code
3. Update to Google Cloud SDK v358 (and ensure that the appengine components update to 1.9.94)
4. Run the sample test code again, observe the exception it throws
5. Try again with Google Cloud SDK v359, observe the same error (it also contains v1.9.94 of the AppEngine components)

## Specifications

- Version: See discussion above
- Platform: I am running the test suite on MacOS 11 11.5 and 11.6

## Root Cause

The testbed runs a sub-process that implements/wraps API calls to the Datastore Emulator, launched as `api_server.py`. The 500 error above occurs when the main process attempts to send a request to the sub-process.

By patching the code that launches the api_server.py within the testbed's code to print STDOUT and STDERR from the api_server.py when it exits, I was able to detect the following stack trace that likely caused the 500 response:

```
Traceback (most recent call last):
  File "/Users/wsorvis/google-cloud-sdk/platform/google_appengine/lib/cherrypy/cherrypy/wsgiserver/wsgiserver2.py", line 1302, in communicate
    req.respond()
  File "/Users/wsorvis/google-cloud-sdk/platform/google_appengine/lib/cherrypy/cherrypy/wsgiserver/wsgiserver2.py", line 831, in respond
    self.server.gateway(self).respond()
  File "/Users/wsorvis/google-cloud-sdk/platform/google_appengine/lib/cherrypy/cherrypy/wsgiserver/wsgiserver2.py", line 2115, in respond
    response = self.req.server.wsgi_app(self.env, self.start_response)
  File "/Users/wsorvis/google-cloud-sdk/platform/google_appengine/google/appengine/tools/devappserver2/wsgi_server.py", line 302, in __call__
    return app(environ, start_response)
  File "/Users/wsorvis/google-cloud-sdk/platform/google_appengine/google/appengine/tools/devappserver2/wsgi_server.py", line 342, in __call__
    return self.app(environ, start_response)
  File "<string>", line 547, in __call__
  File "<string>", line 521, in _handle_CLEAR
  File "/Users/wsorvis/google-cloud-sdk/platform/google_appengine/google/appengine/tools/devappserver2/datastore_grpc_stub.py", line 126, in Clear
    response = six.moves.urllib.urlopen(
AttributeError: 'Module_six_moves_urllib' object has no attribute 'urlopen'
```

The following snippet from `google/appengine/tools/devappserver2/datastore_grpc_stub.py` shows where the error occurred:

```python
  def Clear(self):
    # api_server.py has _handle_CLEAR() method which requires this interface for
    # reusing api_server between unittests.
    response = six.moves.urllib.urlopen(
        six.moves.urllib.Request(
            'http://%s/reset' % self.grpc_apiserver_host, data=''))
    if response.code != six.moves.http_client.OK:
      raise IOError('The Cloud Datastore emulator did not reset successfully.')
```

`six.moves.urllib.urlopen` and `six.moves.urllib.Request` are not valid calls to six, which caused the 500 error. They should be `six.moves.urllib.request.urlopen` and `six.moves.urllib.request.Request`.

By grepping through the legacy runtime's platform files in Google Cloud SDK v358, I spotted a few other cases where it looks like `six` is not being used correctly:

![bad_six_references](https://user-images.githubusercontent.com/62613/135317478-60ad7a54-426f-4775-b361-4cb201598350.png)

## Workaround

For now, we are working around this by pinning to Google Cloud SDK v357.0 when we run tests or run the application locally and in CI.

**Comment by asriniva** [comment]
Thanks for the full root cause analysis! We're tracking this issue internally and are working on a fix.

**Comment by asriniva** [comment]
This has been fixed internally, and should be available with version v1.9.95 of the App Engine components when that releases.
